### ISSUE TITLE
Lockscreen crash

### ISSUE THREAD

**Issue by castrik** [florisboard/florisboard#1632]
#### Short description

I was in lockscreen, putting in my password for the first time after reboot
thankfully, keyboard reappeared after crashing once, so I was able to get in my device

#### Steps to reproduce
1. Go to Lockscreen
2. Type Password
3. See crash

#### Environment information
- FlorisBoard Version: 0.3.14-beta14 (70)
- Device: Samsung SM-G960F
- Android: 10 (cn=Q sdk=29)

#### Features enabled
```kt
smartbar = true
suggestions = true
suggestions_clipboard = true
suggestions_next_word = true
glide = true
clipboard_internal = true
clipboard_history = true
```

#### Attached stacktrace files
<details>
<summary>1646942485172.stacktrace</summary>

```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1465)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier$Pruner.<init>(StatisticalGlideTypingClassifier.kt:403)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.initializePruner(StatisticalGlideTypingClassifier.kt:149)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.setLayout(StatisticalGlideTypingClassifier.kt:117)
	at dev.patrickgold.florisboard.ime.text.gestures.GlideTypingManager.setLayout(GlideTypingManager.kt:67)
	at dev.patrickgold.florisboard.ime.text.keyboard.TextKeyboardLayoutKt.TextKeyboardLayout(TextKeyboardLayout.kt:126)
	at dev.patrickgold.florisboard.ime.text.TextInputLayoutKt.TextInputLayout(TextInputLayout.kt:46)
	at dev.patrickgold.florisboard.ime.text.TextInputLayoutKt$TextInputLayout$2.invoke(Unknown Source:8)
	at dev.patrickgold.florisboard.ime.text.TextInputLayoutKt$TextInputLayout$2.invoke(Unknown Source:8)
	at androidx.compose.runtime.RecomposeScopeImpl.compose(RecomposeScopeImpl.kt:140)
	at androidx.compose.runtime.ComposerImpl.recomposeToGroupEnd(Composer.kt:2158)
	at androidx.compose.runtime.ComposerImpl.skipCurrentGroup(Composer.kt:2404)
	at androidx.compose.runtime.ComposerImpl$doCompose$2$5.invoke(Composer.kt:2585)
	at androidx.compose.runtime.ComposerImpl$doCompose$2$5.invoke(Composer.kt:2571)
	at androidx.compose.runtime.SnapshotStateKt__DerivedStateKt.observeDerivedStateRecalculations(DerivedState.kt:247)
	at androidx.compose.runtime.SnapshotStateKt.observeDerivedStateRecalculations(Unknown Source:1)
	at androidx.compose.runtime.ComposerImpl.doCompose(Composer.kt:2571)
	at androidx.compose.runtime.ComposerImpl.recompose$runtime_release(Composer.kt:2547)
	at androidx.compose.runtime.CompositionImpl.recompose(Composition.kt:620)
	at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:786)
	at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:105)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$2.invoke(Recomposer.kt:456)
	at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$2.invoke(Recomposer.kt:425)
	at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:34)
	at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:109)
	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:997)
	at android.view.Choreographer.doCallbacks(Choreographer.java:797)
	at android.view.Choreographer.doFrame(Choreographer.java:728)
	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:984)
	at android.os.Handler.handleCallback(Handler.java:883)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:237)
	at android.app.ActivityThread.main(ActivityThread.java:8167)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)

```
</details>

**Comment by patrickgold** [comment]
Above linked PR should fix the issue. Will be released in 0.3.14-rc03 today!

**Comment by tsiflimagas** [comment]
I'm getting this again, and somewhat frequently. I was wondering whether it's better to open a new issue, but I'm not really sure. Please let me know.
Also, I think killing FlorisBoard, clearing cache and launching it, can consistently reproduce the crash.

#### Environment information
- FlorisBoard Version: 0.3.15-rc01 (78)
- Device: Xiaomi Redmi Note 8
- Android: 12 (cn=S_V2 sdk=32)

#### Features enabled
```kt
smartbar = true
suggestions = false
suggestions_clipboard = true
suggestions_next_word = true
glide = false
clipboard_internal = false
clipboard_history = false
```

#### Attached stacktrace files
<details>
<summary>1649460807584.stacktrace</summary>

```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1441)
	at java.util.HashMap$KeyIterator.next(HashMap.java:1465)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier$Pruner.<init>(StatisticalGlideTypingClassifier.kt:404)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.initializePruner(StatisticalGlideTypingClassifier.kt:149)
	at dev.patrickgold.florisboard.ime.text.gestures.StatisticalGlideTypingClassifier.setWordData(StatisticalGlideTypingClassifier.kt:132)
	at dev.patrickgold.florisboard.ime.text.gestures.GlideTypingManager$setWordData$1.invokeSuspend(GlideTypingManager.kt:83)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)

```
</details>

**Comment by patrickgold** [comment]
You are indeed correct, I can fully reproduce this bug:

First, switch to any other keyboard. Then kill FlorisBoard. Lock your device, then upon unlocking in the password field click on the keyboard switch icon and select FlorisBoard. See crash. Luckily it recovers and then you can enter the password.

Will investigate this now.

**Comment by patrickgold** [comment]
Above commit should be a fix for the crash, at least I could not reproduce it anymore. Could you please check out the debug artifact to test it on your own device? Thanks in advance!

**Comment by tsiflimagas** [comment]
I tried both mentioned ways, a few times each, and it seems I can't reproduce it either. Great job once again!

**Comment by patrickgold** [comment]
Perfect, thanks for trying it out!

I think the reason why it might have resurfaced is the fact that this crash did only occur if your active subtype was not the first of the list, then it loaded the word data twice at once and both threads accessed the same map at once, which results in this crash. Now there's a lock on it and one thread must wait on the other.
